pkg/sqlite/configmap_writer.go

```diff
@@ -0,0 +1,212 @@
+package sqlite
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+
+	"github.com/ghodss/yaml"
+	_ "github.com/mattn/go-sqlite3"
+	"github.com/sirupsen/logrus"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/operator-framework/operator-registry/pkg/client"
+)
+
+// configmap keys can contain underscores, but configmap names can not
+var unallowedKeyChars = regexp.MustCompile("[^-A-Za-z0-9_.]")
+
+const (
+	ResourceKey                 = "operators.coreos.com.bundle.resources"
+	MediaTypeKey                = "operators.coreos.com.bundle.mediatype"
+	EnvContainerImage           = "CONTAINER_IMAGE"
+	ConfigMapImageAnnotationKey = "olm.sourceImage"
+)
+
+type AnnotationsFile struct {
+	Annotations struct {
+		Resources string `json:"operators.coreos.com.bundle.resources"`
+		MediaType string `json:"operators.coreos.com.bundle.mediatype"`
+	} `json:"annotations"`
+}
+
+type ConfigMapWriter struct {
+	manifestsDir  string
+	configMapName string
+	namespace     string
+	clientset     *kubernetes.Clientset
+}
+
+var _ SQLPopulator = &ConfigMapWriter{}
+
+func NewConfigMapLoaderForDirectory(configMapName, namespace, manifestsDir, kubeconfig string) *ConfigMapWriter {
+	clientset, err := client.NewKubeClient(kubeconfig, logrus.StandardLogger())
+	if err != nil {
+		logrus.Fatalf("cluster config failed: %v", err)
+	}
+
+	return &ConfigMapWriter{
+		manifestsDir:  manifestsDir,
+		configMapName: configMapName,
+		namespace:     namespace,
+		clientset:     clientset,
+	}
+}
+
+func TranslateInvalidChars(input string) string {
+	validConfigMapKey := unallowedKeyChars.ReplaceAllString(input, "~")
+	return validConfigMapKey
+}
+
+func (c *ConfigMapWriter) Populate() error {
+	subDirs := []string{"/manifests/", "/metadata/"}
+
+	configMapPopulate, err := c.clientset.CoreV1().ConfigMaps(c.namespace).Get(c.configMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	configMapPopulate.Data = map[string]string{}
+
+	var totalSize uint64
+	for _, dir := range subDirs {
+		completePath := c.manifestsDir + dir
+		files, err := ioutil.ReadDir(completePath)
+		if err != nil {
+			return err
+		}
+
+		for _, file := range files {
+			log := logrus.WithField("file", completePath+file.Name())
+			log.Info("Reading file")
+			content, err := ioutil.ReadFile(completePath + file.Name())
+			if err != nil {
+				log.Errorf("read failed: %v", err)
+				return err
+			}
+			totalSize += uint64(len(content))
+			if totalSize > (1 << 20) {
+				// over 1 MiB limit, bail out
+				log.Errorf("File with size %v exceeded 1 MiB limit, aboring", len(content))
+				return fmt.Errorf("file %v bigger than total allowed limit", file.Name())
```

Comment:
:+1:

Reply:
Do we want to explain to users why there is a limit?

Reply:
I thought about that, but then I saw that I wasn't 100% sure of the current limit. It seems in the past the etcd storage limit was 1 MB, but now it might be 1.5 MB? https://github.com/etcd-io/etcd/blob/master/Documentation/dev-guide/limit.md#request-size-limit. Personally this was an addition made at the last minute to handle what I consider to be a very unlikely scenario. If we're worried about file size limit I'm not sure the current design is good with (currently just one) configmap backing.

Reply:
Maybe this should be an input on the tool, that defaults to 1mb? Then ask kube changes we can configure it differently? 

(this is also technically a per-cluster limit and can be configured in etcd, and things like k3s don't use etcd could potentially have a different limit - but most clusters probably use the same defaults)

Reply:
Modified as suggested, new parameter added with a default of 1 MB.